File: output/js/Medium/843-most-frequent-subtree-sum.js
/**https://leetcode.com/problems/most-frequent-subtree-sum/ */
//Given the root of a binary tree, return the most frequent subtree sum. If there is a tie, return all the values with the highest frequency in any order.
//The subtree sum of a node is defined as the sum of all the node values formed by the subtree rooted at that node (including the node itself).
// 
//Example 1:
//Input: root = [5,2,-3]
//Output: [2,-3,4]
//Example 2:
//Input: root = [5,2,-5]
//Output: [2]
// 
//Constraints:
//	The number of nodes in the tree is in the range [1, 104].
//	-105 <= Node.val <= 105
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var findFrequentTreeSum = function(root) {
    let map = {};
    let max = -Infinity;
    let result = [];
    let postOrder = (node) => {
        if (!node) return 0;
        let left = postOrder(node.left);
        let right = postOrder(node.right);
        let sum = left + right + node.val;
        map[sum] = map[sum] + 1 || 1;
        max = Math.max(max, map[sum]);
        return sum;
    };
    postOrder(root);
    for (let key in map) {
        if (map[key] === max) result.push(key);
    }
    return result;
};

File: output/py/Medium/843-most-frequent-subtree-sum.py
#https://leetcode.com/problems/most-frequent-subtree-sum/
#Given the root of a binary tree, return the most frequent subtree sum. If there is a tie, return all the values with the highest frequency in any order.
#The subtree sum of a node is defined as the sum of all the node values formed by the subtree rooted at that node (including the node itself).
# 
#Example 1:
#Input: root = [5,2,-3]
#Output: [2,-3,4]
#Example 2:
#Input: root = [5,2,-5]
#Output: [2]
# 
#Constraints:
#	The number of nodes in the tree is in the range [1, 104].
#	-105 <= Node.val <= 105
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def findFrequentTreeSum(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        def dfs(node):
            if node is None:
                return 0

            left = dfs(node.left)
            right = dfs(node.right)
            total = node.val + left + right
            count[total] += 1
            return total

        count = collections.Counter()
        dfs(root)
        maxCount = max(count.values())
        return [total for total in count if count[total] == maxCount]
